src/transport/CertificateAuthenticatedSessionEstablishment.h

```diff
@@ -0,0 +1,171 @@
+#include <transport/raw/PeerAddress.h>
+
+namespace chip {
+
+using namespace Crypto;
+
+class DLL_EXPORT CertificateAuthenticatedSessionEstablishment
+{
+public:
+    CertificateAuthenticatedSessionEstablishment();
+    CertificateAuthenticatedSessionEstablishment(CertificateAuthenticatedSessionEstablishment &&)      = default;
+    CertificateAuthenticatedSessionEstablishment(const CertificateAuthenticatedSessionEstablishment &) = default;
+    CertificateAuthenticatedSessionEstablishment & operator=(const CertificateAuthenticatedSessionEstablishment &) = default;
+    CertificateAuthenticatedSessionEstablishment & operator=(CertificateAuthenticatedSessionEstablishment &&) = default;
+
+    virtual ~CertificateAuthenticatedSessionEstablishment();
+
+    /**
+     * @brief
+     *   Initialize using operational credentials code and wait for session establishment requests.
+     *
+     * @param myNodeId        Node id of local node
+     * @param myKeyId         Key ID to be assigned to the secure session on the peer node
+     * @param delegate        Callback object
+     *
+     * @return CHIP_ERROR     The result of initialization
+     */
+    CHIP_ERROR WaitForSessionEstablishment(NodeId myNodeId, uint16_t myKeyId, AuthenticatedSessionEstablishmentDelegate * delegate);
+
+    /**
+     * @brief
+     *   Create and send session establishment request using device's operational credentials.
+     *
+     * @param peerAddress      Address of peer with which to establish a session.
+     * @param myNodeId         Node id of local node
+     * @param peerNodeId       Node id of the peer node
+     * @param myKeyId          Key ID to be assigned to the secure session on the peer node
+     * @param delegate         Callback object
+     *
+     * @return CHIP_ERROR      The result of initialization
+     */
+    CHIP_ERROR EstablishSession(const Transport::PeerAddress peerAddress, NodeId myNodeId, NodeId peerNodeId, uint16_t myKeyId,
+                                AuthenticatedSessionEstablishmentDelegate * delegate);
+
+    /**
+     * @brief
+     *   Derive a secure session from the established session. The API will return error
+     *   if called before session is established.
+     *
+     * @param session     Reference to the secure session that will be
+     *                    initialized once session establishment is complete
+     * @return CHIP_ERROR The result of session derivation
+     */
+    virtual CHIP_ERROR DeriveSecureSession(SecureSession & session);
+
+    /**
+     * @brief
+     *   Handler for peer's messages, exchanged during pairing handshake.
+     *
+     * @param packetHeader Message header for the received message
+     * @param peerAddress  Source of the message
+     * @param msg          Message sent by the peer
+     * @return CHIP_ERROR The result of message processing
+     */
+    virtual CHIP_ERROR HandlePeerMessage(const PacketHeader & packetHeader, const Transport::PeerAddress & peerAddress,
```

Comment:
Note that the coding conventions called for avoiding `virtual` to save flash cost of vtables on embedded devices.

My understanding is these are here to improve testing via Mock classes.